### ISSUE TITLE
dis_optimizer is not used

### ISSUE THREAD

**Issue by haithemb** [parthagrawal02/MAE_GAN#1]
It seems that the discriminator is not updated with a separate optimizer as in GANs training loop

**Comment by parthagrawal02** [comment]
The paper used Shared Weights for both Discriminator and Generator.

**Comment by haithemb** [comment]
Thank you for the quick reply. My understanding of the paper that it is using "shared weights of generator and discriminator backbones" (3.3. Training Scheme).
The generator backbone (mae encoder) and discriminator without the final linear layer have shared parameters. But the MAE decoder and the linear discriminator layer parameters are not shared.
In algorithm 1 of the paper, it is stated "GAN-MAE model with generator parameters
θmae and discriminator parameters θdisc"
Otherwise I don't think it can be considered adversarial training

**Comment by parthagrawal02** [comment]
That's right. My understanding was that since disc Loss does not include any compute from decoder parameters, their gradients will remain zero, effectively updating only Discriminator parameters. Please correct me if I am wrong.

**Comment by haithemb** [comment]
I think disc Loss does include compute from decoder parameters, since the generated patches are coming from the decoder. So the decoder parameters need to be frozen when updating the discriminator parameters

**Comment by parthagrawal02** [comment]
Sorry, just realised that. Let's freeze the decoder parameters. Thanks for your input.

**Comment by haithemb** [comment]
Many thanks for the quick reply and sharing an updated code

**Comment by parthagrawal02** [comment]
Sorry, wont be able to update the changes, and test right now, will do so soon.

**Comment by haithemb** [comment]
Understood. Thanks

**Comment by GuoyaoShen** [comment]
> I think disc Loss does include compute from decoder parameters, since the generated patches are coming from the decoder. So the decoder parameters need to be frozen when updating the discriminator parameters

Will .detach().clone() work for this? Use this for currupt_img in the model forward function so that the discriminator loss does not keep any conputaion graph for the encoder-decoder in MAE. And since the discriminator is sharing weights with the encoder, it can handle automatically.

I haven't test this though.
